src/main/java/loqor/ait/core/item/BaseGunItem.java

```diff
@@ -0,0 +1,204 @@
+import net.minecraft.world.World;
+
+import loqor.ait.AITMod;
+import loqor.ait.core.AITItems;
+import loqor.ait.core.AITSounds;
+import loqor.ait.core.AITStatusEffects;
+
+public class BaseGunItem extends RangedWeaponItem {
+    public static final Identifier SHOOT = new Identifier(AITMod.MOD_ID, "shoot_gun");
+    public static final Predicate<ItemStack> GUN_PROJECTILES = itemStack -> itemStack.isOf(AITItems.STASER_BOLT_MAGAZINE);
+    public static final double MAX_AMMO = 2000;
+    public static final String AMMO_KEY = "ammo";
+
+    public BaseGunItem(Settings settings) {
+        super(settings);
+    }
+
+    static {
+        ServerPlayNetworking.registerGlobalReceiver(SHOOT, (server, player, handler, buf, responseSender) -> {
+        boolean shoot = buf.readBoolean();
+        boolean isAds = buf.readBoolean();
+
+        if (shoot) {
+            if (player.getMainHandStack().getItem() instanceof BaseGunItem gun) {
+                if (gun.getCurrentAmmo(player.getMainHandStack()) <= 0) {
+                    player.getWorld().playSound(null, player.getX(), player.getY(), player.getZ(), SoundEvents.BLOCK_STONE_BUTTON_CLICK_OFF, SoundCategory.PLAYERS, 1.0f, 1.0f);
+                    return;
+                }
+                BaseGunItem.shoot(player.getWorld(), player, Hand.MAIN_HAND, player.getMainHandStack(), AITItems.STASER_BOLT_MAGAZINE.getDefaultStack(),
+                        1.0f, false, 4.0f, player.hasStatusEffect(AITStatusEffects.ZEITON_HIGH) ? 20f : gun.getAimDeviation(isAds), 0.0f);
+                NbtCompound compound = player.getMainHandStack().getOrCreateNbt();
+                double current = compound.getDouble(AMMO_KEY);
+                double removableAmmo = (isAds ? 15 : 10);
+                if (current - removableAmmo <= 0) {
+                    compound.putDouble(AMMO_KEY, 0);
+                } else {
+                    compound.putDouble(AMMO_KEY, current - removableAmmo <= 0 ? 0 : current - removableAmmo);
+                }
+            }
+        }
+        });
+    }
+
+    @Override
+    public ItemStack getDefaultStack() {
+        ItemStack stack = new ItemStack(this);
+        NbtCompound nbt = stack.getOrCreateNbt();
+        nbt.putDouble(AMMO_KEY, 0);
+        return stack;
+    }
+
+    public static void shootGun(boolean shoot, boolean isAds) {
+        PacketByteBuf buf = PacketByteBufs.create();
+        buf.writeBoolean(shoot);
+        buf.writeBoolean(isAds);
+        ClientPlayNetworking.send(BaseGunItem.SHOOT, buf);
+    }
+
+    @Override
+    public void inventoryTick(ItemStack stack, World world, Entity entity, int slot, boolean selected) {
+        super.inventoryTick(stack, world, entity, slot, selected);
+        if (world.isClient() && entity instanceof PlayerEntity) {
+            if (selected) {
+                BaseGunItem.shootGun(MinecraftClient.getInstance().options.attackKey.isPressed(), MinecraftClient.getInstance().options.useKey.isPressed());
```

Comment:
same as above
move this code into a seperate method, eg `tryShoot`, and annotate that method with `OnlyIn`